src/lein_monolith/task/fingerprint.clj

```diff
@@ -0,0 +1,352 @@
+  (:require
+    [clojure.data]
+    [clojure.edn :as edn]
+    [clojure.java.io :as jio]
+    [clojure.set :as set]
+    [clojure.string :as str]
+    [leiningen.core.main :as lein]
+    [leiningen.core.project :as project]
+    [lein-monolith.dependency :as dep]
+    [lein-monolith.target :as target]
+    [lein-monolith.task.util :as u]
+    [multihash.core :as mhash]
+    [multihash.digest :as digest]
+    [puget.color.ansi :as ansi]
+    [puget.printer :as puget])
+  (:import
+    (java.io
+      File
+      PushbackInputStream)))
+
+
+;; ## Options
+
+(def selection-opts
+  (assoc target/selection-opts :upstream 0 :downstream 0))
+
+
+;; ## Hashing projects' inputs
+
+(def ^:private ->multihash
+  "Globally storing the algorithm we use to generate each multihash."
+  digest/sha1)
+
+
+(defn- aggregate-hashes
+  "Takes a collection of multihashes, and aggregates them together into a unified hash."
+  [mhashes]
+  ;; TODO: is there a better way to do this?
+  (if (= 1 (count mhashes))
+    (first mhashes)
+    (->> mhashes
+         (map mhash/base58)
+         (sort)
+         (apply str)
+         (->multihash))))
+
+
+(defn- list-all-files
+  [^File file]
+  (if (.isFile file)
+    [file]
+    (mapcat list-all-files (.listFiles file))))
+
+
+(defn- hash-file
+  "Takes a File object, and returns a multihash that uniquely identifies the
+  content of this file and the location of the file."
+  [^File file]
+  ;; TODO: only prefix the file location relative to the subproject root?
+  (let [prefix (.getBytes (str (.getAbsolutePath file) "\n"))]
+    (with-open [in (PushbackInputStream. (jio/input-stream file) (count prefix))]
+      (.unread in prefix)
+      (->multihash in))))
+
+
+(defn- all-paths
+  "Finds all source, test, and resource paths associated with a project, including
+  those set in profiles."
+  [project]
+  (->> (concat
+         [project]
+         (vals (:profiles project)))
+       (mapcat (juxt :source-paths :test-paths :resource-paths))
+       (mapcat identity)
+       (map (fn absolute-file
+              [dir-str]
+              ;; Monolith subprojects and profiles don't use absolute paths
+              (if (str/starts-with? dir-str (:root project))
+                (jio/file dir-str)
+                (jio/file (:root project) dir-str))))))
+
+
+(defn- hash-sources
+  [project]
+  (->> (all-paths project)
+       (mapcat list-all-files)
+       (map hash-file)
+       (aggregate-hashes)))
+
+
+(defn- hash-dependencies
+  [project]
+  (-> (:dependencies project)
+      (pr-str)
+      (->multihash)))
+
+
+(declare hash-inputs)
+
+
+(defn- hash-upstream-projects
+  [project dep-map subprojects cache]
+  (->> (dep-map (dep/project-name project))
+       (keep subprojects)
+       (map #(::final (hash-inputs % dep-map subprojects cache)))
+       (aggregate-hashes)))
+
+
+(defn- cache-result!
+  [cache project m]
+  (get (swap! cache assoc (dep/project-name project) m) (dep/project-name project)))
```

Comment:
This seems more complicated than returning `m` directly, since you're not modifying it based on an existing entry.